pkg/cloudinfo/alibaba/api_response.go

```diff
@@ -0,0 +1,72 @@
+// Copyright © 2019 Banzai Cloud
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package alibaba
+
+import (
+	"github.com/aliyun/alibaba-cloud-sdk-go/sdk/requests"
+	"github.com/aliyun/alibaba-cloud-sdk-go/sdk/responses"
+	"github.com/banzaicloud/cloudinfo/pkg/cloudinfo"
+	"github.com/pkg/errors"
+	"strconv"
+	"time"
+)
+
+// CommonDescriber interface for all Alibaba API calls
+type CommonDescriber interface {
+	ProcessCommonRequest(request *requests.CommonRequest) (*responses.CommonResponse, error)
+}
+
+func (a *AlibabaInfoer) apiResponse(name, region string, instanceTypes []string) (*responses.CommonResponse, error) {
+	request := requests.NewCommonRequest()
+	request.Method = "POST"
+	request.ApiName = name
+	switch name {
+	case "DescribeSpotPriceHistory":
+		request.Domain = "ecs.aliyuncs.com"
+		request.Version = "2014-05-26"
+		request.QueryParams["RegionId"] = region
+		request.QueryParams["NetworkType"] = "vpc"
+		request.QueryParams["OSType"] = "linux"
+		request.QueryParams["StartTime"] = time.Now().Round(1 * time.Hour).UTC().Format(time.RFC3339)
+		request.QueryParams["InstanceType"] = instanceTypes[0]
+	case "DescribeInstanceTypes":
+		request.Domain = "ecs.aliyuncs.com"
+		request.Version = "2014-05-26"
+	case "DescribeZones":
+		request.Domain = "ecs.aliyuncs.com"
+		request.Version = "2014-05-26"
+		request.QueryParams["RegionId"] = region
+	case "GetPayAsYouGoPrice":
+		request.Scheme = "https"
+		request.Domain = "business.ap-southeast-1.aliyuncs.com"
+		request.Version = "2017-12-14"
+		request.QueryParams["RegionId"] = region
+		request.QueryParams["ProductCode"] = "ecs"
+		request.QueryParams["SubscriptionType"] = "PayAsYouGo"
+		for i, instanceType := range instanceTypes {
+			request.QueryParams[cloudinfo.CreateString("ModuleList.", strconv.Itoa(i+1), ".ModuleCode")] = "InstanceType"
+			request.QueryParams[cloudinfo.CreateString("ModuleList.", strconv.Itoa(i+1), ".Config")] =
+				cloudinfo.CreateString("InstanceType:", instanceType, ",IoOptimized:IoOptimized,ImageOs:linux")
+			request.QueryParams[cloudinfo.CreateString("ModuleList.", strconv.Itoa(i+1), ".PriceType")] = "Hour"
+		}
+	case "DescribeRegions":
+		request.Domain = "ecs.aliyuncs.com"
+		request.Version = "2014-05-26"
+		request.QueryParams["AcceptLanguage"] = "en-US"
+	default:
+		return nil, errors.New("invalid API name")
+	}
+	return a.client.ProcessCommonRequest(request)
```

Comment:
let the client call as a separate method / tfactory methods are meant to assemble different requests the client executes the api calls (this is a kind of "separation of responsibilities")